Trying to Decode E164 Address Type But its failing

Hi ,
Can some one help us to understand why the decode for Address Family is not happening as expected. 

We have SMSC-Address , which is of Address Family and of type E164, when we define "Address" in our dictionary and try to Decode this it adds the Type  value at the front of the parsed value as this  "\x00\b436990071301" instead of "436990071301".

![SMSC_Addres](https://github.com/user-attachments/assets/ff023a6d-e74c-4e02-ab09-a5c509a0683e)


Example Parsed RAW value: 

SMSC-Address {Code:2017,Flags:0xc0,Length:28,VendorId:10415,Value:Address{datatype.Address{0x34, 0x33, 0x36, 0x39, 0x39, 0x30, 0x30, 0x37, 0x31, 0x33, 0x30, 0x31}}, Type{datatype.Address{0x0, 0x8}} Padding:2},

func TestDecodeAddressE164(t *testing.T) {
	b := []byte{0x00, 0x08, 0x34, 0x33, 0x36, 0x39, 0x39, 0x30, 0x30, 0x37, 0x31, 0x33, 0x30, 0x31}
	address, err := DecodeAddress(b)
	if err != nil {
		t.Fatal(err)
	}
	if address.Padding() != 3 {
		t.Fatalf("Unexpected padding. Want 3, have %d",
			address.Padding())
	}
}

Result : --- FAIL: TestDecodeAddressE164 (1.87s)
    address_test.go:206: Unexpected padding. Want 3, have 2
FAIL

--- a/diam/datatype/address.go
+++ b/diam/datatype/address.go
@@ -34,8 +34,11 @@ func DecodeAddress(b []byte) (Type, error) {
 			return Address{}, errors.New("Invalid length for IPv6")
 		}
 	default:
+		// Preserve the full slice (family prefix + value) for unknown
+		// address families so Serialize can round-trip them untouched.
 		return Address(d), nil
 	}
+	// IPv4/IPv6: strip the 2-byte address family prefix.
 	return Address(d[2:]), nil
 }
 
@@ -51,6 +54,7 @@ func (addr Address) Serialize() []byte {
 		b[1] = 0x02
 		copy(b[2:], addr)
 	} else {
+		// Non-IP: addr already contains the family prefix from decode.
 		b = make([]byte, len(addr))
 		copy(b, addr)
 	}
@@ -60,24 +64,16 @@ func (addr Address) Serialize() []byte {
 // Len implements the Type interface.
 func (addr Address) Len() int {
 	if ip4 := net.IP(addr).To4(); ip4 != nil {
-		return len(ip4) + 2 // two bytes from the address family
+		return len(ip4) + 2
 	} else if ip6 := net.IP(addr).To16(); ip6 != nil {
-		return len(ip6) + 2 // two bytes from the address family
-	} else {
-		return len(addr)
+		return len(ip6) + 2
 	}
+	return len(addr) // family prefix is already in the slice
 }
 
 // Padding implements the Type interface.
 func (addr Address) Padding() int {
-	var l int
-	if ip4 := net.IP(addr).To4(); ip4 != nil {
-		l = len(ip4) + 2 // two bytes from the address family
-	} else if ip6 := net.IP(addr).To16(); ip6 != nil {
-		l = len(ip6) + 2 // two bytes from the address family
-	} else {
-		l = len(addr)
-	}
+	l := addr.Len()
 	return pad4(l) - l
 }
 
@@ -94,8 +90,12 @@ func (addr Address) String() string {
 	if ip6 := net.IP(addr).To16(); ip6 != nil {
 		return fmt.Sprintf("Address{%s},Padding:%d", net.IP(addr), addr.Padding())
 	}
-	if len(addr) == 0 {
-		return "Address{},Padding:0" // NOTE: To avoid panicking on addr[2:]
+	if len(addr) < 2 {
+		return fmt.Sprintf("Address{%#v},Padding:%d", []byte(addr), addr.Padding())
 	}
-	return fmt.Sprintf("Address{%#v},Type{%#v},Padding:%d", addr[2:], addr[:2], addr.Padding())
+	// Non-IP (e.g. E.164): show the value as a readable string and the
+	// family as raw bytes. Fixes #201 where decoded E.164 values printed
+	// as "\x00\x08<number>".
+	return fmt.Sprintf("Address{%s},Type{%#v},Padding:%d",
+		string(addr[2:]), []byte(addr[:2]), addr.Padding())
 }

--- a/diam/datatype/address_test.go
+++ b/diam/datatype/address_test.go
@@ -7,6 +7,7 @@ package datatype
 import (
 	"bytes"
 	"net"
+	"strings"
 	"testing"
 )
 
@@ -167,10 +168,11 @@ func TestAddressIPv6_Generic(t *testing.T) {
 }
 
 func TestAddressE164_Generic(t *testing.T) {
-	var addressBytes []byte
-	addressType := []byte{0x0, 0x8}
+	// E.164 Address keeps the 2-byte family prefix (0x0008) in the slice
+	// so it can round-trip back to the wire verbatim.
+	addressType := []byte{0x00, 0x08}
 	addressValue := []byte("48602007060")
-	addressBytes = make([]byte, len(addressType)+len(addressValue))
+	addressBytes := make([]byte, len(addressType)+len(addressValue))
 	copy(addressBytes[:2], addressType)
 	copy(addressBytes[2:], addressValue)
 	address := Address(addressBytes)
@@ -201,9 +203,28 @@ func TestDecodeAddressE164(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
+	// The family prefix is kept in the slice; the value is after it.
+	addr := address.(Address)
+	if len(addr) < 2 {
+		t.Fatalf("Address too short: %d bytes", len(addr))
+	}
+	if got := string(addr[2:]); got != "48602007060" {
+		t.Fatalf("Unexpected value. Want 48602007060, have %q", got)
+	}
 	if address.Padding() != 3 {
 		t.Fatalf("Unexpected padding. Want 3, have %d",
 			address.Padding())
+	}
+	if address.Len() != 13 {
+		t.Fatalf("Unexpected len. Want 13, have %d", address.Len())
+	}
+	// Round-trip: serialize should produce the original wire bytes.
+	if v := address.Serialize(); !bytes.Equal(v, b) {
+		t.Fatalf("Round-trip failed. Want 0x%x, have 0x%x", b, v)
+	}
+	// #201: String() must not embed raw family bytes inside the value.
+	if s := address.String(); strings.Contains(s, "\x00\x08") {
+		t.Fatalf("String() leaks raw family bytes into output: %q", s)
 	}
 }
 

